### PR TITLE
allow onlyoffice docker and integrated

### LINKED ISSUE #1275: Make it possible to choose the Docker or Nextcloud App (OnlyOffice Collabora)
@szaimen Same thing with Collabora now that happened to OnlyOffice, I think we should move in the same way as we did with OnlyOffice.

Basically, copy the OnlyOffice script, change to Collabora and then move the old script to `/old`.

### PR DESCRIPTION
**Pull request by szaimen**
attempt to fix #1275

### CODE REVIEW

**Comment by szaimen** [comment]
I haven't tested this but I think it is already a good foundation to see how the feature could be achieved.

File: apps/collabora.sh [hunk 1]
```diff
@@ -23,6 +23,16 @@ lowest_compatible_nc 13
 ram_check 2 Collabora
 cpu_check 2 Collabora
 
+# Check if Nextcloud is installed with TLS
+if ! occ_command_no_check config:system:get overwrite.cli.url | grep -q "https"
+then
+msg_box "Sorry, but Nextcloud needs to be run on HTTPS which doesn't seem to be the case here.
+You easily activate TLS (HTTPS) by running the Let's Encrypt script found in $SCRIPTS.
```
Comment (enoch85) [inline apps/collabora.sh:30]: This should refer to menu.sh.

File: apps/collabora.sh [hunk 2]
```diff
@@ -23,6 +23,16 @@ lowest_compatible_nc 13
...
+More info here: https://bit.ly/37wRCin
+To run this script again, just exectue 'sudo bash $SCRIPTS/menu.sh' and choose Collabora."
```
Comment (enoch85) [inline apps/collabora.sh:32]: choose Additional Apps --> Collabora

**Comment by enoch85** [comment]
What I was thinking was an entry in Additional Apps which says "Online Document Server" and then another menu with all the different options:

- Collabora Intergrated
- OnlyOffice Integrated
- Collabora Docker
- OnlyOffice Docker

It would be cleaner not to clutter the Additional Apps even more.

**Comment by szaimen** [comment]
> What I was thinking was an entry in Additional Apps which says "Online Document Server" and then another menu with all the different options:
> 
> * Collabora Intergrated
> * OnlyOffice Integrated
> * Collabora Docker
> * OnlyOffice Docker
> 
> It would be cleaner not to clutter the Additional Apps even more.

Sounds good 👍

**Comment by szaimen** [comment]
Shall that be a menu that gets triggered from additional apps or be another option in main_menu?

**Comment by szaimen** [comment]
I ask because network_okay resets the network adapter every time which is really slow and putting more scripts into scripts will slow down the experience...

**Comment by szaimen** [comment]
I guess what makes it so slow is the countdown here:
https://github.com/nextcloud/vm/blob/2bfd76b73bae39969a6b3dc53e9eae690b60ab86/lib.sh#L775
Can we remove that or move it into the if-statement above where the network gets resettet?

**Comment by enoch85** [comment]
> Shall that be a menu that gets triggered from additional apps or be another option in main_menu?

Please make it Menu --> Apps --> new menu called Documentserver

Thanks!

File: menu/main_menu.sh [hunk 3]
```diff
@@ -25,33 +25,21 @@ choice=$(whiptail --title "Main Menu" --radiolist "Choose what you want to do.\n
...
         then
             bash $SCRIPTS/update.sh
         else
-            if network_ok
-            then
-                download_script STATIC update
-                bash $SCRIPTS/update.sh
-            fi
+            download_script STATIC update
+            bash $SCRIPTS/update.sh
```
Comment (szaimen) [inline menu/main_menu.sh:42]: Just if you wonder: I've made that change because to get into this script we will have already done a network check with should be enough, imho...

Comment (enoch85) [reply menu/main_menu.sh]: IMHO, we don't need the network checks at all.

Comment (szaimen) [reply menu/main_menu.sh]: I removed them 👍

File: lib.sh [hunk 4]
```diff
@@ -770,9 +770,9 @@ else
     if ! netplan apply
     then
         systemctl restart systemd-networkd > /dev/null
+        countdown 'Waiting for network to restart...' 3
     fi
     # Check the connention
-    countdown 'Waiting for network to restart...' 3
```
Comment (enoch85) [inline lib.sh:775]: Why this change?

Comment (szaimen) [reply lib.sh:775]: Because I though this would be only needed if we actually restart the network in the if-statement above?
Or does netplan apply actually restart the network so that we have to wait for this 3 seconds?

Comment (enoch85) [reply lib.sh:775]: > Or does netplan apply actually restart the network so that we have to wait for this 3 seconds?

It should check if the network is OK even after netplan apply, so yes, it's needed where it was from the beginning. :)

Comment (szaimen) [reply lib.sh:775]: Hm, I don't get it. I just moved that line two lines higher, I didn't remove it...

Comment (enoch85) [reply lib.sh:775]: Yeah, and now you don't get if network is online if `netplan apply` exits with 0, which is wrong - because it should check after `netplan apply` as well. :)

Comment (szaimen) [reply lib.sh:775]: But the countdown function doesn't check if the network is online, does it?

Comment (enoch85) [reply lib.sh:775]: No, but right now it doesn't wait for `netplan apply` either, which it should.

Comment (szaimen) [reply lib.sh]: I've reverted that change.

File: apps/collabora_docker.sh [hunk 5]
```diff
@@ -23,7 +23,19 @@ lowest_compatible_nc 13
...
+msg_box "Sorry, but Nextcloud needs to be run on HTTPS which doesn't seem to be the case here.
+You easily activate TLS (HTTPS) by running the Let's Encrypt script.
+More info here: https://bit.ly/37wRCin
+
+To run this script again, just exectue 'sudo bash $SCRIPTS/menu.sh' and choose:
+Additional Apps --> Collabora."
+    exit
+fi
+
+# Check if Collabora is already installed
```
Comment (enoch85) [inline apps/collabora_docker.sh:38]: Since we use this in multiple places, please make it a function.

Maybe something like `check_nextcloud_https`?

File: menu/additional_apps.sh [hunk 6]
```diff
@@ -18,12 +18,11 @@ root_check
 
 # Install Apps
 choice=$(whiptail --title "Which apps do you want to install?" --checklist "Automatically configure and install selected apps\nSelect by pressing the spacebar\nYou can view this menu later by running 'sudo bash $SCRIPTS/menu.sh'" "$WT_HEIGHT" "$WT_WIDTH" 4 \
+"Documentserver" "(Choose between Onlyoffice, Collabora and Docker, integrated)" OFF \
```
Comment (enoch85) [inline menu/additional_apps.sh:21]: (OnlyOffice or Collabora - Docker or Integrated)

File: menu/additional_apps.sh [hunk 7]
```diff
@@ -34,6 +33,11 @@ choice=$(whiptail --title "Which apps do you want to install?" --checklist "Auto
 "SMB-mount" "(Connect to SMB-shares from your local network)" OFF 3>&1 1>&2 2>&3)
 
 case "$choice" in
+    *"Documentserver"*)
+        clear
+        print_text_in_color "$ICyan" "Downloading Documentserver.sh..."
```
Comment (enoch85) [inline menu/additional_apps.sh:38]: As I mentioned before, I think it looks better if the info says "Downloading the X script..."

Please replace in all places.

File: menu/documentserver.sh [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+choice=$(whiptail --title "Which Documentserver do you want to install?" --radiolist "Automatically configure and install the selected Documentserver.\nSelect by pressing the spacebar and ENTER\nYou can view this menu later by running 'sudo bash $SCRIPTS/menu.sh'" "$WT_HEIGHT" "$WT_WIDTH" 4 \
+"Collabora (Docker)" "(Online editing [2GB RAM] - subdomain required)" OFF \
+"Collabora (integrated)" "(Online editing [2GB RAM] - no subdomain required)" OFF \
+"OnlyOffice (Docker)" "(Online editing [2GB RAM] - subdomain required)" OFF \
+"OnlyOffice (integrated)" "(Online editing - no subdomain required)" OFF 3>&1 1>&2 2>&3)
+
+case "$choice" in
+    "Collabora (Docker)")
+        clear
+        print_text_in_color "$ICyan" "Downloading Collabora_Docker.sh..."
```
Comment (enoch85) [inline menu/documentserver.sh:28]: Downloading script for X...

**Comment by enoch85** [comment]
Great job here! Kudos! :rocket:

File: menu/documentserver.sh [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        clear
+        print_text_in_color "$ICyan" "Downloading Collabora_integrated.sh..."
+        run_script APP collabora_integrated
+    ;;
+    "OnlyOffice (Docker)")
+        clear
+        print_text_in_color "$ICyan" "Downloading OnlyOffice_Docker.sh..."
+        run_script APP onlyoffice_docker
+    ;;
+    "OnlyOffice (integrated)")
```
Comment (enoch85) [inline menu/documentserver.sh:41]: (Integrated)

Please use capital "I" for all occurrences, everywhere.

Thanks!

File: apps/collabora_docker.sh (hunk 5 above)
Comment (szaimen) [reply apps/collabora_docker.sh:29]: done

File: menu/additional_apps.sh (hunk 6 above)
Comment (szaimen) [reply menu/additional_apps.sh]: done

File: menu/additional_apps.sh (hunk 7 above)
Comment (szaimen) [reply menu/additional_apps.sh]: done

File: menu/documentserver.sh (hunk 8 above)
Comment (szaimen) [reply menu/documentserver.sh]: kind of done?

File: menu/documentserver.sh (hunk 9 above)
Comment (szaimen) [reply menu/documentserver.sh:41]: done

File: apps/collabora_docker.sh [hunk 10]
```diff
@@ -227,6 +230,12 @@ then
   SSLProxyVerify None
   SSLProxyCheckPeerCN Off
   SSLProxyCheckPeerName Off
+  
+  # Improve security settings
+  Header set X-XSS-Protection "1; mode=block"
+  Header set Strict-Transport-Security "max-age=31536000; includeSubDomains; preload"
+  Header set X-Content-Type-Options nosniff
+  Header set Content-Security-Policy "frame-ancestors 'self' ${NCDOMAIN//\\/}"
```
Comment (szaimen) [inline apps/collabora_docker.sh:238]: I'll add this comment so that we don't forget to test this.

File: lib.sh [hunk 11]
```diff
@@ -498,6 +498,19 @@ then
...
 
+# Check if Nextcloud is installed with TLS
+check_nextcloud_https() {
+    if ! occ_command_no_check config:system:get overwrite.cli.url | grep -q "https"
+    then
+    msg_box "Sorry, but Nextcloud needs to be run on HTTPS which doesn't seem to be the case here.
+    You easily activate TLS (HTTPS) by running the Let's Encrypt script.
+    More info here: https://bit.ly/37wRCin
+    To run this script again, just exectue 'sudo bash $SCRIPTS/menu.sh' and choose:
+    Additional Apps --> Documentserver --> $@."
```
Comment (enoch85) [inline lib.sh]: Please use $1 and $2 is possible. Using @ is risky.

Thanks!

File: apps/collabora_docker.sh (hunk 10 above)
Comment (enoch85) [reply apps/collabora_docker.sh:238]: Except this, is everything else tested?

Comment (enoch85) [reply apps/collabora_docker.sh:238]: Just had a quick look, menus looks good. Haven't tested the scripts but I guess you copied from old (working) stuff?

Comment (szaimen) [reply apps/collabora_docker.sh:238]: Yes, I copied most things from old (working) stuff 👍

File: lib.sh (hunk 11 above)
Comment (enoch85) [reply lib.sh]: ping @szaimen

File: menu/additional_apps.sh [hunk 12]
```diff
@@ -36,8 +36,12 @@ case "$choice" in
         clear
         print_text_in_color "$ICyan" "Downloading the Documentserver script..."
         run_script MENU documentserver
+   *"Bitwarden "*)
+        clear
+        print_text_in_color "$ICyan" "Downloading the Bitwarden script..."
+        run_script MENU bitwarden_menu
     ;;&
-    *"Fail2ban"*)
+    *"Fail2ban "*)
```
Comment (enoch85) [inline menu/additional_apps.sh:44]: needs fixing

**Comment by enoch85** [comment]
Fixed the conflicts, please check that everything is in order.

File: lib.sh (hunk 11 above)
Comment (szaimen) [reply lib.sh]: this was already changed...

File: menu/additional_apps.sh (hunk 12 above)
Comment (szaimen) [reply menu/additional_apps.sh]: This was actually by design. because if you don't have the space in there, if you choose Fail2ban-Statuscheck, it will run also the Fail2Ban script...

Comment (enoch85) [reply menu/additional_apps.sh]: Aah, ok, please revert.

Comment (szaimen) [reply menu/additional_apps.sh:45]: okay, is reverted

**Comment by szaimen** [comment]
> Fixed the conflicts, please check that everything is in order.

Looks still good, you just changed something in additional_apps and merged the master into this branch, if I am not mistaken...

**Comment by szaimen** [comment]
Master has still an issue? (look at travis)

**Comment by enoch85** [comment]
> Looks still good, you just changed something in additional_apps and merged the master into this branch, if I am not mistaken...

This branch wasn't even with master, so after I merged the Bitwarden menu stuff, I needed to update here as well. It was a bit tricky but I think I managed.

**Comment by szaimen** [comment]
> This branch wasn't even with master, so after I merged the Bitwarden menu stuff, I needed to update here as well. It was a bit tricky but I think I managed.

Yah, I think it worked 👍

BTW: is there a new issue in master? (new travis issue)

**Comment by enoch85** [comment]
> BTW: is there a new issue in master? (new travis issue)

Thanks, fixed now. :+1:

**Comment by enoch85** [comment]
Please let me know when you've tested this. Thanks!

**Comment by enoch85** [comment]
@szaimen I want to merge this tonight. Have you tested it?

**Comment by enoch85** [comment]
Setting up a VPS now, will do some basic tests before merge.

**Comment by enoch85** [comment]
Documentserver script works
Collabora Docker works
OnlyOffice Docker works
Collabora Integrated works (need to manually change DocumentServer in the GUI settings though)
OnlyOffice Integrated doesn't work. Will check that.

**Comment by enoch85** [comment]
OnlyOffice won't install correctly...

**Review by enoch85 — APPROVED** [review]

**Comment by szaimen** [comment]
🎉

**Comment by szaimen** [comment]
@enoch85 I would actually change that back to 
`"OnlyOffice (Integrated)" "(Online editing - no subdomain required)"`
https://github.com/nextcloud/vm/blob/f00d35644efbf48939ef778dc22311037e4c8052/menu/documentserver.sh#L23
Since in the script itself is no cpu- and ram-check (and also not needed, afaik) and no requirement for a subdomain...

**Comment by enoch85** [comment]
Oops, copy paste mistake. :)

**Comment by szaimen** [comment]
Shall I create a PR or do you handle this?

**Comment by enoch85** [comment]
Please create one, otherwise I will forget. :)
